Content.Server/Weapons/Ranged/Systems/GunSystem.Battery.cs

```diff
@@ -36,12 +40,33 @@ private void OnBatteryChargeChange(EntityUid uid, BatteryAmmoProviderComponent c
         UpdateShots(uid, component, args.Charge, args.MaxCharge);
     }
 
+    private void OnPowerCellChanged(EntityUid uid, BatteryAmmoProviderComponent component, ref PowerCellChangedEvent args)
+    {
+        UpdateShots(uid, component);
+    }
+
+    private void UpdateShots(Entity<BatteryAmmoProviderComponent> entity)
+    {
+        UpdateShots(entity.Owner, entity.Comp);
+    }
+
     private void UpdateShots(EntityUid uid, BatteryAmmoProviderComponent component)
     {
-        if (!TryComp<BatteryComponent>(uid, out var battery))
-            return;
+        var currentCharge = 0f;
+        var maxCharge = 0f;
+        if (TryComp<BatteryComponent>(uid, out var battery))
+        {
+            currentCharge = battery.CurrentCharge;
+            maxCharge = battery.MaxCharge;
+        }
+        else if (TryComp<PowerCellSlotComponent>(uid, out var powerCellSlot) &&
+            _powerCell.TryGetBatteryFromSlot(uid, out var powerCell, powerCellSlot))
+        {
+            currentCharge = powerCell.CurrentCharge;
+            maxCharge = powerCell.MaxCharge;
+        }
 
-        UpdateShots(uid, component, battery.CurrentCharge, battery.MaxCharge);
+        UpdateShots(uid, component, currentCharge, maxCharge);
     }
```

Comment:
Isn't the whole point of chargechanged to handle this? As is this is turning into a bit of a mess trying to handle 2 codepaths and if we start having to do this everywhere it will soon become unmaintainable.